File: leetcode/c++/easy/_1399_count_largest_group.cpp
// Leetcode Problem 1399

#include <gtest/gtest.h>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int countLargestGroup(int n) {
        unordered_map<int, int> groups;

        int sum;
        for (int i = 1; i <= n; i++) {
            sum = 0;
            int temp = i;
            while (temp > 0) {
                sum += temp % 10;
                temp /= 10;
            }
            groups[sum]++;
        }

        int max_occurances = 0;
        for (auto group = groups.begin(); group != groups.end(); ++group) {
            if (group->second > max_occurances) {
                max_occurances = group->second;
            }
        }
        int count = 0;
        for (auto group = groups.begin(); group != groups.end(); ++group) {
            if (group->second == max_occurances) {
                count++;
            }
        }

        return count;
    }
};


class TestCountLargestGroup : public ::testing::Test {
protected:
    Solution solution;
};

TEST_F(TestCountLargestGroup, SimpleTestCase1) {
    EXPECT_EQ(solution.countLargestGroup(13), 4);
    EXPECT_EQ(solution.countLargestGroup(2), 2);
    EXPECT_EQ(solution.countLargestGroup(99), 1);
}

File: leetcode/rust/problems/src/easy/_1399_count_largest_group.rs
// Leetcode Problem 1399

use std::collections::HashMap;

struct Solution;

#[allow(dead_code)]
impl Solution {
    pub fn count_largest_group(n: i32) -> i32 {
        let mut groups: HashMap<i32, i32> = HashMap::new();
        
        let mut sum;
        for i in 1..=n {
            sum = 0;
            let mut temp = i;
            while temp > 0 {
                sum += temp % 10;
                temp /= 10;
            }
            groups.entry(sum).and_modify(|e| *e += 1).or_insert(1);
        }

        if let Some(&max_occurance) = groups.values().max() {
            groups.values().filter(|&&group| group == max_occurance).count() as i32
        } else {
            0
        }
    }
}


#[test]
fn test_count_largest_group() {
    assert_eq!(Solution::count_largest_group(13), 4);
    assert_eq!(Solution::count_largest_group(2), 2);
    assert_eq!(Solution::count_largest_group(16), 7);
    assert_eq!(Solution::count_largest_group(99), 1);
}